tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_scanner.c

```diff
@@ -0,0 +1,221 @@
+}
+
+/** Bypassing setting up transmission, and will try to send the raw data that is
+ *  provided to the function.
+ */
+static void test_tx_send_ad_type_msg(uint8_t type, uint8_t *data, uint8_t len)
+{
+	struct bt_le_adv_param param = {};
+	uint16_t duration, adv_int;
+	struct bt_data ad;
+	int err;
+
+	adv_int = ADV_INT_FAST_MS;
+	duration = ADV_DURATION;
+
+	ad.type = type;
+	ad.data_len = len;
+	ad.data = data;
+
+	param.id = BT_ID_DEFAULT;
+	param.interval_min = BT_MESH_ADV_SCAN_UNIT(adv_int);
+	param.interval_max = param.interval_min;
+
+	LOG_DBG("ad.type %u len %u", ad.type, ad.data_len);
+	LOG_HEXDUMP_DBG(ad.data, ad.data_len, "ad.data");
+
+	uint64_t st = k_uptime_get();
+
+	err = bt_le_adv_start(&param, &ad, 1, NULL, 0);
+	if (err) {
+		FAIL("Advertising failed: err %d", err);
+		return;
+	}
+	LOG_DBG("Advertising started. Sleeping %u ms", duration);
+
+	k_sleep(K_MSEC(duration));
+	err = bt_le_adv_stop();
+	if (err) {
+		FAIL("Stopping advertising failed: err %d", err);
+		return;
+	}
+	LOG_DBG("Advertising stopped (%u ms)", (uint32_t) k_uptime_delta(&st));
+}
+
+/** Test sending message with not supported ad type for mesh.
+ *
+ * First message send message with an ad data type not supported by mesh, and
+ * verify that the receiver is disregarding message. Then send the same message
+ * with correct ad type to verify that the message is received.
+ */
+static void test_tx_invalid_ad_type(void)
+{
+	test_tx_device_setup();
+
+	/* Create a valid message */
+	uint8_t data[28] = {
+		0x0d, 0x10, 0xca, 0x54, 0xd0,
+		0x00, 0x24, 0x00, 0xaa, 0x8c,
+		0xcc, 0x6b, 0x6a, 0xc8, 0x51,
+		0x69, 0x16, 0x4d, 0xf6, 0x9b,
+		0xce, 0xbd, 0xc7, 0xa3, 0xf0,
+		0x28, 0xdf, 0xae
+	};
+	uint8_t data_len = ARRAY_SIZE(data);
+
+	LOG_DBG("TX Invalid AD Type");
+
+	k_sleep(K_SECONDS(1));
+
+	/* Send message with invalid ad type. */
+	test_tx_send_ad_type_msg(BT_DATA_BIG_INFO, data, data_len);
+
+	/* Wait for no message receive window to end. */
+	k_sleep(K_SECONDS(10));
+
+	/* Send message with valid ad type to verify message. */
+	test_tx_send_ad_type_msg(BT_DATA_MESH_MESSAGE, data, data_len);
+
+	PASS();
+}
+
+/** Test sending message with invalid/valid ad type and wrong packet length.
+ *
+ * Send messages with wrong packet length, and verify that the receiver is
+ * disregarding message. Then send the same message with correct packet length
+ * to verify that the message is received.
+ */
+static void test_tx_wrong_packet_length(void)
+{
+	test_tx_device_setup();
+
+	/* Create a valid message */
+	uint8_t data[28] = {
+		0x0d, 0x10, 0xca, 0x54, 0xd0,
+		0x00, 0x24, 0x00, 0xaa, 0x8c,
+		0xcc, 0x6b, 0x6a, 0xc8, 0x51,
+		0x69, 0x16, 0x4d, 0xf6, 0x9b,
+		0xce, 0xbd, 0xc7, 0xa3, 0xf0,
+		0x28, 0xdf, 0xae
+	};
+	uint8_t data_len = ARRAY_SIZE(data);
+
+	LOG_DBG("TX wrong packet length");
+
+	k_sleep(K_SECONDS(1));
+
+	/* Send message with to long data length. */
+	test_tx_send_ad_type_msg(BT_DATA_MESH_MESSAGE, data, data_len+1);
```

Comment:
This'll read bytes that are out of bounds, no? Perhaps the valid message should be padded, e.g. to max adv len?

Reply:
We could pad the message, but I do not see any point in doing that. The last byte will then just be a random value. We can not add than 1, because if we add 2 it will fail with message about that the adv message is to large and it will not send the message.